### PR TITLE
[DomCrawler] Attach label to form fields

### PR DESCRIPTION
**Pull request by carlosV2**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

While building a software to extract information from a webpage, I need to fill a form with hints on the fields labels.

Example:

``` html
<form>
   ...
   <label for="my_field">Input the result of 3 + 4:</label>
   <input type="text" id="my_field" name="my_field" />
   ...
</form>
```

 I thought it would be handy to have the field labels attached to each field.
So given the previous form you could now do:

``` php
$form->get('my_field')->getLabel()->textContent;
```

### CODE REVIEW

File: src/Symfony/Component/DomCrawler/Form.php [hunk 1]
```diff
@@ -454,22 +454,49 @@ private function addField(\DOMElement $node)
...
     }
+
+    /**
+     * Finds the associated label to a given field node.
+     *
+     * @param \DOMElement $node The \DOMElement instance of the field
+     *
+     * @return \DOMNode|null The \DOMElement of the associated label tag or null if none was found
+     */
+    private function findAssociatedLabel(\DOMElement $node)
```
Comment (jakzal) [inline src/Symfony/Component/DomCrawler/Form.php]: Have you considered moving this implementation to the FormField's `getLabel()` method? You wouldn't have to add a new argument to FormField's constructor, and the logic wouldn't be called if the label wasn't accessed.  `ownerDocument` property is available on the `$node` that's already passed to the FormField.

Comment (carlosV2) [inline src/Symfony/Component/DomCrawler/Form.php]: That's a good point. I'll refactor it.

File: src/Symfony/Component/DomCrawler/Field/FormField.php [hunk 2]
```diff
@@ -46,7 +46,7 @@
     /**
      * Constructor.
      *
-     * @param \DOMElement $node The node associated with this field
+     * @param \DOMElement $node  The node associated with this field
```
Comment (stof) [inline src/Symfony/Component/DomCrawler/Field/FormField.php]: should be reverted

**Comment by carlosV2** [comment]
Sorry @stof, I realised while you where commenting it. It should be fixed now.

File: src/Symfony/Component/DomCrawler/Field/FormField.php [hunk 3]
```diff
@@ -58,6 +58,30 @@ public function __construct(\DOMElement $node)
...
+    public function getLabel()
+    {
+        $xpath = new \DOMXPath($this->node->ownerDocument);
+
+        $labels = $xpath->query('ancestor::label[1]', $this->node);
+        if ($labels->length > 0) {
+            return $labels->item(0);
+        }
+
+        if ($this->node->hasAttribute('id')) {
```
Comment (stof) [inline src/Symfony/Component/DomCrawler/Field/FormField.php]: I think the id-based logic should have higher priority

**Comment by xabbuh** [comment]
What about something like this:

```html
```

**Comment by carlosV2** [comment]
@xabbuh it will work as far as the label is within the form node computed by the crawler. In this case, parenting would fail but the label will still be found by the `for` attribute.

If it is suitable, I can add a new test to prove this although it seems to be working on my local.

**Comment by xabbuh** [comment]
@carlosV2 You are right. I misread the way the XPath object is created.

**Comment by fabpot** [comment]
Thank you @carlosV2.

**Pull request closed without merging**